### DISCUSSION TITLE
Hammerspoon self-reloading not working

### DISCUSSION THREAD

**Discussion by unphased** [Hammerspoon/hammerspoon#3672] [Q&A] [answered]
I get this intermittently functioning, which means there is a race condition involved: 

    -- watcher object must be assigned to global not to get GC'd
    myWatcher = hs.pathwatcher.new(os.getenv("HOME") .. "/.vim/hammerspoon-init.lua/", function(files) hs.reload() end):start()

My hammerspoon init lives under ~/.vim/, and this gets symlinked into ~:

```
❯ ll ~/.hammerspoon/
total 0
lrwxr-xr-x    1 slu  staff    36B Aug 12 20:32 init.lua -> /Users/slu/.vim/hammerspoon-init.lua
```

Anyone else seen similar problems? How to mitigate the race condition? When it fails, I get this notif: 
<img width="364" alt="image" src="https://github.com/user-attachments/assets/9312aaf4-df42-4a21-b5e4-d2a5613119a9">

Console reports: 

```
2024-08-15 23:09:47: -- Lazy extension loading enabled
2024-08-15 23:09:47: -- Loading extension: fnutils
2024-08-15 23:09:47: -- Can't find ~/.hammerspoon/init.lua; create it and reload your config.
```

I found this https://github.com/Hammerspoon/hammerspoon/issues/2478 which may be related but it's hard to tell, plus, that thread indicates whatever issues were there got resolved.

**Comment by unphased** [comment]
I changed the code to: 

    myWatcher = hs.pathwatcher.new(os.getenv("HOME") .. "/.vim/hammerspoon-init.lua/", hs.reload):start()

~~and it started working reliably again. I'll come back and close this once it works reliably for a few more days.~~

No dice, same problem. (clicking reload on the hammerspoon tray icon menu always loads it properly)

**Reply by unphased** [reply]
I'm thinking it could be a race condition vis-a-vis neovim/vim file save shenanigans. for whatever weird reason vim does a whole dance when it saves files, so hammerspoon probably is getting the event for the file a bit early and tries to read it at a moment when the file isn't readable.

**Reply by latenitefilms** [reply]
Maybe try wrapping in a `hs.timer`?

**Reply by unphased** [reply, marked as answer]
code now: 

```lua
myWatcher = hs.pathwatcher.new(os.getenv("HOME") .. "/.vim/hammerspoon-init.lua/", function(files)
    -- Use a timer to delay the reload slightly
    hs.timer.doAfter(0.5, function()
        hs.reload()
    end)
end):start()
```

Seems good so far.
